### ISSUE TITLE
Codepoint support in regular expressions

### ISSUE THREAD

**Issue by jmitchell** [tree-sitter/tree-sitter#286]
# Problem

Tree-sitter refuses to parse a `grammar.js` with a regular expression which ranges over codepoints:

```
$ cat grammar.js
module.exports = grammar({
  name: 'unicode',

  extras: $ => [],

  rules: {
    source_file: $ => /[\u{20}-\u{10FFFF}]/u
  }
});
$ tree-sitter generate
Rule source_file_token1 regex parse error:
    [\u{20}-\u\{10FFFF\}]
              ^
error: invalid hexadecimal digit
```

If I instead use the regular expression `/[\u20-\u10FFFF]/` the `tree-sitter generate` step succeeds, but the resulting grammar doesn't actually parse all codepoints between U+20 and U+10FFFF, inclusive.

# Background

While working on a `grammar.js` for Dhall I discovered `→` (U+2192) isn't recognized by `/[\u20-\u10FFFF]/`, yet that same regex recognizes `π` (U+03C0). Initially I was surprised both codepoints are clearly within the specified range.

After some digging I learned this regular expression deals with surrogates instead of codepoints as I had expected (see `RegExpUnicodeEscapeSequence` in the ECMAScript spec). To refer to codepoints I need to use curly braces and a `u` regex flag: `/[\u{20}-\u{10FFFF}]/u`. This works as expected with JavaScript, but as mentioned not yet with tree-sitter.

```
> '\u03C0'.match(/[\u{20}-\u{10FFFF}]/u)
[ 'π', index: 0, input: 'π' ]
> '\u2192'.match(/[\u{20}-\u{10FFFF}]/u)
[ '→', index: 0, input: '→' ]
```

More details are captured at https://github.com/jmitchell/abnf-to-tree-sitter/issues/1.

**Comment by jmitchell** [comment]
I hadn't tried converting `\u{10FFFF}` into the corresponding surrogate pair yet. When I do tree-sitter panics.

```
$ cat grammar.js
module.exports = grammar({
  name: 'unicode',

  extras: $ => [],

  rules: {
    source_file: $ => /[\u0020-\uDBFF\uDFFF]/
  }
});
$ RUST_BACKTRACE=1 tree-sitter generate
thread 'main' panicked at 'called `Option::unwrap()` on a `None` value', src/libcore/option.rs:355:21
stack backtrace:
   0: std::sys::unix::backtrace::tracing::imp::unwind_backtrace
             at src/libstd/sys/unix/backtrace/tracing/gcc_s.rs:49
   1: std::sys_common::backtrace::_print
             at src/libstd/sys_common/backtrace.rs:71
   2: std::panicking::default_hook::{{closure}}
             at src/libstd/sys_common/backtrace.rs:59
             at src/libstd/panicking.rs:211
   3: std::panicking::default_hook
             at src/libstd/panicking.rs:227
   4: std::panicking::rust_panic_with_hook
             at src/libstd/panicking.rs:491
   5: std::panicking::continue_panic_fmt
             at src/libstd/panicking.rs:398
   6: rust_begin_unwind
             at src/libstd/panicking.rs:325
   7: core::panicking::panic_fmt
             at src/libcore/panicking.rs:95
   8: core::panicking::panic
             at src/libcore/panicking.rs:59
   9: tree_sitter_cli::generate::nfa::CharacterSet::add_range
  10: tree_sitter_cli::generate::prepare_grammar::expand_tokens::NfaBuilder::expand_character_class
  11: tree_sitter_cli::generate::prepare_grammar::expand_tokens::NfaBuilder::expand_regex
  12: tree_sitter_cli::generate::prepare_grammar::expand_tokens::NfaBuilder::expand_rule
  13: tree_sitter_cli::generate::prepare_grammar::expand_tokens::expand_tokens
  14: tree_sitter_cli::generate::prepare_grammar::prepare_grammar
  15: tree_sitter_cli::generate::generate_parser_for_grammar_with_opts
  16: tree_sitter_cli::generate::generate_parser_in_directory
  17: tree_sitter::main
  18: std::rt::lang_start::{{closure}}
  19: std::panicking::try::do_call
             at src/libstd/rt.rs:59
             at src/libstd/panicking.rs:310
  20: __rust_maybe_catch_panic
             at src/libpanic_unwind/lib.rs:102
  21: std::rt::lang_start_internal
             at src/libstd/panicking.rs:289
             at src/libstd/panic.rs:398
             at src/libstd/rt.rs:58
  22: main
  23: __libc_start_main
  24: _start
```

That helps us know where to look!

**Comment by jmitchell** [comment]
This patch to `dsl.js` works around the initial regex parse error:

```diff
diff --git a/cli/src/generate/dsl.js b/cli/src/generate/dsl.js
index c18ac530..a477c93a 100644
--- a/cli/src/generate/dsl.js
+++ b/cli/src/generate/dsl.js
@@ -1,4 +1,5 @@
 const UNICODE_ESCAPE_PATTERN = /\\u([0-9a-f]{4})/gi;
+const UNICODE_CODEPOINT_ESCAPE_PATTERN = /\\u\{([0-9a-f]+)\}/gi;
 
 function alias(rule, value) {
   const result = {
@@ -178,6 +179,10 @@ function normalize(value) {
               UNICODE_ESCAPE_PATTERN,
               (match, group) => String.fromCharCode(parseInt(group, 16))
             )
+           .replace(           // TODO: only use when input RegExp has 'u' flag.
+             UNICODE_CODEPOINT_ESCAPE_PATTERN,
+             (match, group) => String.fromCodePoint(parseInt(group, 16))
+           )
       };
     case ReferenceError:
       throw value
```

Now the original test `grammar.js` hits the same panic error shown in the preceding post. The failing `unwrap` must come from [calling `char::from_u32`](https://github.com/tree-sitter/tree-sitter/blob/858b4ba8accc648edbe414576038b116b0c0f805/cli/src/generate/nfa.rs#L80) on a non-scalar unicode code point.

After making the following change and trying to generate the original test grammar again the panic included the message `'char::from_u32(d800) failed'`.

```diff
diff --git a/cli/src/generate/nfa.rs b/cli/src/generate/nfa.rs                                                                                                                    
index ca2e5405..87da0b17 100644
--- a/cli/src/generate/nfa.rs
+++ b/cli/src/generate/nfa.rs
@@ -77,7 +77,7 @@ impl CharacterSet {
         if let CharacterSet::Include(mut chars) = self {
             let mut c = start as u32;
             while c <= end as u32 {
-                chars.push(char::from_u32(c).unwrap());
+                chars.push(char::from_u32(c).expect(&format!("char::from_u32({:x}) failed", c)));
                 c += 1;
             }
             chars.sort_unstable();
```

`\u{D800}` is a leading surrogate, which [causes Rust to barf](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=23513b57940ac5ca4dd582705cd757ef).

> `help: unicode escape must not be a surrogate`

More from Rust's [undefined behavior](https://doc.rust-lang.org/reference/behavior-considered-undefined.html):

> * Invalid values in primitive types, even in private fields and locals:
>     * ...
>     * A value in a char which is a surrogate or above char::MAX.

Of course I don't want to venture into undefined behavior so I'm avoiding [`std::char::from_u32_unchecked`](https://doc.rust-lang.org/std/char/fn.from_u32_unchecked.html).

I suppose the next step is to grok how `CharacterSet`s in Rust compile to C and consider whether there's a safe way to build a range of unicode codepoints without introducing undefined behavior or affecting the C API.

**Comment by maxbrunsfeld** [comment]
Hmm, I'm not exactly sure where the problem is.

Tree-sitter already deals exclusively with full Unicode code points (e.g. Rust's `char` type). This is different from JavaScript, in which strings are made up of 16-bit "code units". So I don't think we should need to make any changes to `CharacterSet` - that type already represents a set of Unicode code points.

It seems like a regex parsing/preprocessing issue. The `\u{abcd}` syntax might be JavaScript-specific; the rust `regex_syntax` library might not understand it.

**Comment by CAD97** [comment]
I think this was fixed by #319.
